optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py

```diff
@@ -0,0 +1,326 @@
+        order = args.cit_order_of_combinations
+        cit = Cit(parameters, order)
+        self.headers, self.variants = cit.combine()
+
+    @staticmethod
+    def error_exit(args):
+        if args.subcommand == 'run':
+            sys.exit(exit_codes.AVOCADO_JOB_FAIL)
+        else:
+            sys.exit(exit_codes.AVOCADO_FAIL)
+
+    def __iter__(self):
+        if self.variants is None:
+            return
+
+        variant_ids = []
+        for variant in self.variants:
+            variant_ids.append("-".join([variant.get(key)
+                                         for key in self.headers]))
+
+        for vid, variant in zip(variant_ids, self.variants):
+            variant_tree_nodes = []
+            for key, val in variant.items():
+                variant_tree_nodes.append(TreeNode(key, {key: val}))
+
+            yield {"variant_id": vid,
+                   "variant": variant_tree_nodes,
+                   "paths": self.parameter_path}
+
+    def __len__(self):
+        return sum(1 for _ in self.variants) if self.variants else 0
+
+    def update_defaults(self, defaults):
+        pass
+
+    def to_str(self, summary, variants, **kwargs):
+        """
+        Return human readable representation
+
+        The summary/variants accepts verbosity where 0 means silent and
+        maximum is up to the plugin.
+
+        :param summary: How verbose summary to output (int)
+        :param variants: How verbose list of variants to output (int)
+        :param kwargs: Other free-form arguments
+        :rtype: str
+        """
+        if not self.variants:
+            return ""
+
+        out = []
+        verbose = variants > 1
+        out.append("CIT Variants (%i):" % len(self))
+        for variant in self:
+            out.append('%sVariant %s:    %s' % ('\n' if verbose else '',
+                                                variant["variant_id"],
+                                                self.parameter_path))
+            if not verbose:
+                continue
+            env = set()
+            for node in variant["variant"]:
+                for key, value in iteritems(node.environment):
+                    origin = node.environment.origin[key].path
+                    env.add(("%s:%s" % (origin, key), str(value)))
+            if not env:
+                return out
+            fmt = '    %%-%ds => %%s' % max([len(_[0]) for _ in env])
+            for record in sorted(env):
+                out.append(fmt % record)
+
+        return "\n".join(out)
+
+
+class Cit(object):
+
+    MATRIX_ROW_SIZE = 20
+    MAX_ITERATIONS = 15
+
+    def __init__(self, parameters, order):
+        # Parameters come as ('key', ['value1', 'value2', 'value3'])
+        self.parameters = parameters
+        # Length (number of values) for each parameter
+        self.parameters_length = [len(param[1]) for param in self.parameters]
+        # Order of combinations
+        self.order = order
```

Comment:
This needs to be `min(order, parameters_length)`, otherwise it reports no results... (`avocado variants --cit-parameter-file examples/varianter_cit/params.ini --cit-order-of-combinations 6`)

Reply:
Yes, we either have to go with `min()` or advise the user that he has given a value that is larger than the number of parameters.  For now, I agree with you that `min()` is the simpler and best option.

But the length we need here is the length of different keys, so it's `len(parameters)`.

Reply:
sure, the result is the same (len(parameters) == len(parameters_length))